### PR TITLE
Add WorkerInfo through TorchBind to make it an available type in TorchScript

### PR DESCRIPTION
**Pull request by xush6528**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#35447 Add WorkerInfo through TorchBind to make it an available type in TorchScript**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D7923053/)

as titled

Differential Revision: [D7923053](https://our.internmc.facebook.com/intern/diff/D7923053/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/rpc_agent.h [hunk 1]
```diff
@@ -36,7 +36,7 @@ struct RpcBackendOptions {
 };
 
 // A globally unique ID to identify an RpcAgent
-struct TORCH_API WorkerInfo {
+struct TORCH_API WorkerInfo : torch::CustomClassHolder {
   WorkerInfo(std::string name, int id)
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rpc_agent.h]: We can remove this constructor?

**Review by mrshenli — COMMENTED** [review]
Changes in RPC LGTM

File: torch/csrc/distributed/rpc/rpc_agent.h (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/distributed/rpc/rpc_agent.h]: agreed, we should keep `WorkerInfo(std::string name, int64_t id)` as the only constructor.

File: torch/csrc/jit/runtime/register_distributed_ops.cpp [hunk 2]
```diff
@@ -63,10 +68,13 @@ RegisterOperators reg_rpc_ops(
...
          aliasAnalysisFromSchema()),
      Operator(
-         "aten::owner(RRef(t) self) -> int",
+         "aten::owner(RRef(t) self) -> __torch__.torch.classes.dist_rpc.WorkerInfo",
          [](Stack& stack) {
            auto rref = pop(stack).toRRef();
-           push(stack, rref->owner());
+           push(
+               stack,
+               torch::make_custom_class<distributed::rpc::WorkerInfo>(
```
Comment (wanchaol) [inline torch/csrc/jit/runtime/register_distributed_ops.cpp]: why we need to make a new instance here? can't we do `RRefContext::getInstance().agent()->getWorkerInfo(rref_->owner());`?

Comment (xush6528) [reply torch/csrc/jit/runtime/register_distributed_ops.cpp]: IValue constructor can take `intrusive_ptr<WorkerInfo>`, but `RRefContext::getInstance().agent()->getWorkerInfo(..)` returns a const ref, `const WorkerInfo&`. 

We don't want it to return a `intrusive_ptr<WorkerInfo>`, because the `allWorkerInfo_` vector is a static setup on constructing RpcAgent, we must let RpcAgent decide it's lifetime.

Similarly, Python side API, `rref.owner()`, asks for this data as well, so we make a copy of this const WorkerInfo, and let Python holds the shared_ptr to this WorkerInfo copy, and Python and decide the lifetime of the copy arbitrarily, without being affected by destruction of RpcAgent and it's contained `allWorkerInfo_` vector.

File: torch/csrc/distributed/rpc/process_group_agent.cpp [hunk 3]
```diff
@@ -135,7 +135,7 @@ ProcessGroupAgent::ProcessGroupAgent(
   }
 
   allWorkerInfo_.reserve(pg_->getSize());
-  for (int rank = 0; rank < (int)tmpWorkerIds.size(); ++rank) {
+  for (worker_id_t rank = 0; rank < (int)tmpWorkerIds.size(); ++rank) {
```
Comment (wanchaol) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: lint failure here.

File: torch/csrc/jit/runtime/register_distributed_ops.cpp (hunk 2 above)
Comment (wanchaol) [reply torch/csrc/jit/runtime/register_distributed_ops.cpp]: I see, that make sense, but when the rpcAgent shutdowns, won't the workerInfos not valid anymore, in this case user is holding a non-valid workerInfo, and user might later pass it with a newly created rpcAgent that holds a different worker list, will that cause problem?

File: test/backward_compatibility/check_backward_compatibility.py [hunk 4]
```diff
@@ -116,7 +116,7 @@
     ('aten::dequantize', datetime.date(2020, 4, 1)),
     ('aten::confirmed_by_owner', datetime.date(2020, 3, 17)),
     ('aten::owner', datetime.date(2020, 3, 27)),
-    ('aten::owner_name', datetime.date(2020, 3, 27)),
+    ('aten::owner_name', datetime.date(2020, 4, 15)),
```
Comment (wanchaol) [inline test/backward_compatibility/check_backward_compatibility.py]: why is this needed? we are changing `aten::owner` instead of `aten::owner_name` right?

File: torch/csrc/distributed/rpc/rpc_agent.h (hunk 1 above)
Comment (xush6528) [reply torch/csrc/distributed/rpc/rpc_agent.h]: removed that constructor.
there are sites that passed int as worker_id, I need to convert them to int64_t to match the int64_t constructor that does an overflow check.

File: test/backward_compatibility/check_backward_compatibility.py (hunk 4 above)
Comment (xush6528) [reply test/backward_compatibility/check_backward_compatibility.py]: Removing it, since no BC test failure is triggered.

I don't know in what case op schema change will be treated as a BC breakage, only if used by production model?

File: torch/csrc/jit/runtime/register_distributed_ops.cpp (hunk 2 above)
Comment (xush6528) [reply torch/csrc/jit/runtime/register_distributed_ops.cpp]: I only image application code holds it for printing debugging related info. e.g. we have rref leak detection code, which could make use of this information.

Your concern is possibly true, since WorkerInfo are just id + name. In new session, the id->name mapping could be different. This will take some effort to fix. 

We didn't prioritize because it's never a real use case. If we will have a new distributed training session, workers are usually new sub-process. Highly impossible to be shared about WorkerInfo copy.

File: torch/csrc/distributed/rpc/process_group_agent.cpp [hunk 5]
```diff
@@ -135,7 +135,7 @@ ProcessGroupAgent::ProcessGroupAgent(
   }
 
   allWorkerInfo_.reserve(pg_->getSize());
-  for (int rank = 0; rank < (int)tmpWorkerIds.size(); ++rank) {
+  for (worker_id_t rank = 0; rank < tmpWorkerIds.size(); ++rank) {
```
Comment (wanchaol) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: can we fix this clang-tidy issue? although I guess the `tmpWorkerIds` is valid when looping, but would like to ensure there's no overflow.

**Review by wanchaol — APPROVED** [review]
Looks good, thanks!

**Comment by mrshenli** [comment]
@xush6528 @wanchaol Do we know why `aten::owner` breaks bc test but `aten:: owner_name` is fine?

https://app.circleci.com/pipelines/github/pytorch/pytorch/147976/workflows/42355ea6-c2b1-4819-86b8-739271148302/jobs/4985127
https://app.circleci.com/pipelines/github/pytorch/pytorch/148738/workflows/0d9a29e0-045a-4922-b5c0-5767217b2f72/jobs/5000992

**Comment by wanchaol** [comment]
> @xush6528 @wanchaol Do we know why `aten::owner` breaks bc test but `aten:: owner_name` is fine?
> 
> [app.circleci.com/pipelines/github/pytorch/pytorch/147976/workflows/42355ea6-c2b1-4819-86b8-739271148302/jobs/4985127](https://app.circleci.com/pipelines/github/pytorch/pytorch/147976/workflows/42355ea6-c2b1-4819-86b8-739271148302/jobs/4985127)
> [app.circleci.com/pipelines/github/pytorch/pytorch/148738/workflows/0d9a29e0-045a-4922-b5c0-5767217b2f72/jobs/5000992](https://app.circleci.com/pipelines/github/pytorch/pytorch/148738/workflows/0d9a29e0-045a-4922-b5c0-5767217b2f72/jobs/5000992)

This PR changes the return type of `aten::owner`, but the exemption date of `aten::ownder` passed, fixed it in https://github.com/pytorch/pytorch/pull/35703
